File: apps/patter-voice-sdk/libraries/typescript/src/providers/gemini-live.ts
/**
 * Gemini Live realtime adapter.
 *
 * Implements Patter's realtime adapter surface — connect / sendAudio /
 * onEvent / close — matching OpenAIRealtimeAdapter.
 *
 * Uses the @google/genai SDK lazily imported at connect() so consumers that do
 * not use Gemini Live do not pay the load cost. Install with:
 *
 *    npm install @google/genai
 *
 * NOTE: Native-audio Gemini Live models are **v1alpha-only**. We pass
 * `httpOptions: { apiVersion: 'v1alpha' }` when constructing the client.
 * When Google promotes native audio to GA, switch to `v1beta` / `v1` and
 * update the default model below.
 * See: https://ai.google.dev/gemini-api/docs/live
 */

import { getLogger } from '../logger';
import { sanitizeGeminiSchema } from './google-llm';

export const GEMINI_DEFAULT_INPUT_SR = 16000;
export const GEMINI_DEFAULT_OUTPUT_SR = 24000;

/** Callback signature for events emitted by {@link GeminiLiveAdapter}. */
export type GeminiLiveEventHandler = (
  type:
    | 'audio'
    | 'transcript_input'
    | 'transcript_output'
    | 'function_call'
    | 'speech_started'
    | 'response_done'
    | 'error',
  data: unknown,
) => void | Promise<void>;

interface GeminiLiveOptions {
  model?: string;
  voice?: string;
  instructions?: string;
  language?: string;
  tools?: Array<{ name: string; description: string; parameters: Record<string, unknown> }>;
  inputSampleRate?: number;
  outputSampleRate?: number;
  temperature?: number;
}

/** Realtime adapter for Google's Gemini Live native-audio API. */
export class GeminiLiveAdapter {
  private readonly model: string;
  private readonly voice: string;
  private readonly instructions: string;
  private readonly language: string;
  private readonly tools?: Array<{ name: string; description: string; parameters: Record<string, unknown> }>;
  private readonly inputSampleRate: number;
  /** Output sample rate — exposed so callers can configure downstream transcoding. */
  readonly outputSampleRate: number;
  private readonly temperature: number;

  private client: unknown = null;
  private session: unknown = null;
  private receiveLoop: Promise<void> | null = null;
  private handlers: GeminiLiveEventHandler[] = [];
  private running = false;
  /**
   * Tracks call_id -> function name so tool responses can be sent back with
   * the correct `name` field (Gemini expects the original function name,
   * not the call_id).
   */
  private pendingToolCalls: Map<string, string> = new Map();

  constructor(
    private readonly apiKey: string,
    options: GeminiLiveOptions = {},
  ) {
    // gemini-2.0-flash-exp was experimental preview retired Dec 2024.
    // gemini-live-2.5-flash-preview was shut down Dec 9, 2025.
    // Current native-audio live model (v1alpha-only) is the dated preview.
    // Callers can override via GeminiLive({ model: ... }).
    // Default model — see https://ai.google.dev/gemini-api/docs/live
    this.model = options.model ?? 'gemini-2.5-flash-native-audio-preview-09-2025';
    this.voice = options.voice ?? 'Puck';
    this.instructions = options.instructions ?? '';
    this.language = options.language ?? 'en-US';
    this.tools = options.tools;
    this.inputSampleRate = options.inputSampleRate ?? GEMINI_DEFAULT_INPUT_SR;
    this.outputSampleRate = options.outputSampleRate ?? GEMINI_DEFAULT_OUTPUT_SR;
    this.temperature = options.temperature ?? 0.8;
  }

  /** Lazily import @google/genai, open a Live session, and start the receive loop. */
  async connect(): Promise<void> {
    let genaiModule: { GoogleGenAI: new (args: { apiKey: string; httpOptions?: Record<string, unknown> }) => unknown };
    try {
      // Lazy dynamic import — keeps @google/genai optional.
      // Variable module name avoids TS2307 when the peer dep is not installed.
      const modName = '@google/genai';
      genaiModule = (await import(modName)) as typeof genaiModule;
    } catch {
      throw new Error(
        '\nGemini Live requires the "@google/genai" package, which is not installed.\n\n' +
          '  Install:  npm install @google/genai\n\n' +
          'This is an optional peer dependency of getpatter — it is only needed when\n' +
          'you use GeminiLive as an agent engine. Other LLM/engine providers do not\n' +
          'require it.\n',
      );
    }

    const { GoogleGenAI } = genaiModule;
    // Native-audio models require the v1alpha endpoint — see module doc.
    this.client = new GoogleGenAI({
      apiKey: this.apiKey,
      httpOptions: { apiVersion: 'v1alpha' },
    });

    const config: Record<string, unknown> = {
      responseModalities: ['AUDIO'],
      speechConfig: {
        voiceConfig: { prebuiltVoiceConfig: { voiceName: this.voice } },
        languageCode: this.language,
      },
      temperature: this.temperature,
      // Without these, native-audio sessions produced NO user transcript
      // ever and no assistant transcript in AUDIO modality — logs/history/
      // metrics got nothing for Gemini Live calls. Mirrors Python.
      inputAudioTranscription: {},
      outputAudioTranscription: {},
    };
    if (this.instructions) {
      config.systemInstruction = { parts: [{ text: this.instructions }] };
    }
    if (this.tools?.length) {
      config.tools = [
        {
          functionDeclarations: this.tools.map((t) => ({
            name: t.name,
            description: t.description,
            // Strip JSON-Schema keys the Live API's proto Schema rejects
            // ($schema, additionalProperties — strict-mode and zod-derived
            // MCP tools): one such tool 400'd the whole session.
            parameters: sanitizeGeminiSchema(t.parameters),
          })),
        },
      ];
    }

    // The genai live surface is organised as client.live.connect({model, config, callbacks?}).
    // Some SDK versions return a Session-like object with send*/receive methods.
    const liveApi = (this.client as { live?: { connect?: (args: unknown) => Promise<unknown> } }).live;
    if (!liveApi?.connect) {
      throw new Error('@google/genai: live.connect is not available in this version');
    }
    this.session = await liveApi.connect({ model: this.model, config });
    this.running = true;

    // Start the receive pump.
    this.receiveLoop = this.pumpReceive().catch((err) => {
      getLogger().error(`Gemini Live receive loop error: ${String(err)}`);
    });
  }

  /** Send a PCM audio chunk to Gemini as base64 inline data. */
  sendAudio(pcm: Buffer): void {
    if (!this.session || !this.running) return;
    const mime = `audio/pcm;rate=${this.inputSampleRate}`;
    const sess = this.session as { sendRealtimeInput?: (args: unknown) => unknown };
    const result = sess.sendRealtimeInput?.({
      media: { data: pcm.toString('base64'), mimeType: mime },
    });
    if (result instanceof Promise) {
      void result.catch((err) =>
        getLogger().warn(`Gemini Live sendAudio error: ${String(err)}`),
      );
    }
  }

  /** Send a text turn to Gemini and mark the turn complete. */
  async sendText(text: string): Promise<void> {
    if (!this.session) return;
    const sess = this.session as { sendClientContent?: (args: unknown) => Promise<void> };
    await sess.sendClientContent?.({
      turns: { role: 'user', parts: [{ text }] },
      turnComplete: true,
    });
  }

  /** Send a tool/function-call result back to Gemini. */
  async sendFunctionResult(callId: string, result: string): Promise<void> {
    if (!this.session) return;
    const sess = this.session as { sendToolResponse?: (args: unknown) => Promise<void> };
    // Gemini requires the original function name in the response, not the
    // call_id. Look it up from the map populated when the tool call was
    // emitted; fall back to callId if we never saw this id (defensive).
    const name = this.pendingToolCalls.get(callId) ?? callId;
    this.pendingToolCalls.delete(callId);
    await sess.sendToolResponse?.({
      functionResponses: [
        { id: callId, name, response: { result } },
      ],
    });
  }

  /** No-op — Gemini Live barge-in is VAD-driven, not client-cancelled. */
  cancelResponse(): void {
    // Gemini Live barge-in is VAD-driven — explicit cancel not in v1alpha wire protocol.
    getLogger().debug('Gemini Live: cancelResponse is implicit via VAD');
  }

  /** Register an event handler that receives every Gemini Live event. */
  onEvent(handler: GeminiLiveEventHandler): void {
    this.handlers.push(handler);
  }

  private async emit(
    type:
      | 'audio'
      | 'transcript_input'
    | 'transcript_output'
      | 'function_call'
      | 'speech_started'
      | 'response_done'
      | 'error',
    data: unknown,
  ): Promise<void> {
    for (const h of this.handlers) {
      try {
        await h(type, data);
      } catch (err) {
        getLogger().error(`Gemini Live handler threw: ${String(err)}`);
      }
    }
  }

  private async pumpReceive(): Promise<void> {
    if (!this.session) return;
    const sess = this.session as { receive?: () => AsyncIterable<unknown> };
    if (typeof sess.receive !== 'function') {
      getLogger().warn('Gemini Live: session.receive() not available');
      return;
    }
    try {
      for await (const response of sess.receive()) {
        if (!this.running) break;
        const r = response as {
          serverContent?: {
            modelTurn?: {
              parts?: Array<{
                inlineData?: { data?: string };
                text?: string;
              }>;
            };
            inputTranscription?: { text?: string };
            outputTranscription?: { text?: string };
            turnComplete?: boolean;
            interrupted?: boolean;
          };
          goAway?: { timeLeft?: string };
          toolCall?: {
            functionCalls?: Array<{
              id?: string;
              name?: string;
              args?: Record<string, unknown> | string;
            }>;
          };
        };

        const sc = r.serverContent;
        if (sc) {
          for (const part of sc.modelTurn?.parts ?? []) {
            if (part.inlineData?.data) {
              await this.emit('audio', Buffer.from(part.inlineData.data, 'base64'));
            }
            if (part.text) await this.emit('transcript_output', part.text);
          }
          if (sc.inputTranscription?.text) {
            await this.emit('transcript_input', sc.inputTranscription.text);
          }
          if (sc.outputTranscription?.text) {
            await this.emit('transcript_output', sc.outputTranscription.text);
          }
          if (sc.turnComplete) await this.emit('response_done', null);
          if (sc.interrupted) await this.emit('speech_started', null);
        }
        if (r.goAway) {
          // Gemini Live hard-caps session length (~10-15 min without
          // resumption); goAway is the only warning before the server drops
          // the connection. Surface it loudly.
          getLogger().warn(
            `Gemini Live goAway received — session ends in ${r.goAway.timeLeft ?? 'unknown'}`,
          );
        }
        if (r.toolCall) {
          for (const fn of r.toolCall.functionCalls ?? []) {
            const args = fn.args ?? {};
            const callId = fn.id ?? '';
            const fnName = fn.name ?? '';
            if (callId && fnName) {
              this.pendingToolCalls.set(callId, fnName);
            }
            await this.emit('function_call', {
              call_id: callId,
              name: fnName,
              arguments: typeof args === 'string' ? args : JSON.stringify(args),
            });
          }
        }
      }
    } catch (err) {
      if (this.running) await this.emit('error', err);
    } finally {
      this.running = false;
    }
  }

  /** Close the Gemini Live session and stop the receive loop. */
  async close(): Promise<void> {
    this.running = false;
    if (this.session) {
      const sess = this.session as { close?: () => Promise<void> | void };
      try {
        await sess.close?.();
      } catch {
        /* ignore */
      }
      this.session = null;
    }
    this.client = null;
    if (this.receiveLoop) {
      await this.receiveLoop.catch(() => undefined);
      this.receiveLoop = null;
    }
    this.pendingToolCalls.clear();
  }
}

File: apps/patter-voice-sdk/libraries/python/getpatter/providers/gemini_live.py
"""Gemini Live realtime adapter.

Patter's adapter surface — ``connect`` / ``send_audio`` / ``receive_events`` /
``close`` — matches :class:`~getpatter.providers.openai_realtime.OpenAIRealtimeAdapter`,
so callers can swap providers without touching the handler. Session lifecycle
(reconnects, resumes) is managed by Patter's handlers, not by the adapter.

NOTE: Native-audio Gemini Live models are **v1alpha-only**. The client must
pass ``http_options={"api_version": "v1alpha"}`` when constructing the genai
client (see :meth:`GeminiLiveAdapter.connect`). When Google promotes native
audio to GA, move the default to ``v1beta`` or ``v1`` and update the default
``model`` below accordingly.
See: https://ai.google.dev/gemini-api/docs/live
"""

from __future__ import annotations

import asyncio
import json
import logging
from enum import IntEnum, StrEnum
from typing import Any, AsyncIterator, Union

logger = logging.getLogger("getpatter.gemini_live")


class GeminiLiveModel(StrEnum):
    """Known Gemini Live (v1alpha) realtime models."""

    NATIVE_AUDIO_PREVIEW_09_2025 = "gemini-2.5-flash-native-audio-preview-09-2025"
    LIVE_2_5_FLASH_PREVIEW = "gemini-live-2.5-flash-preview"
    LIVE_2_0_FLASH_EXP = "gemini-2.0-flash-exp"


class GeminiLiveVoice(StrEnum):
    """Built-in voices accepted by Gemini Live ``PrebuiltVoiceConfig``."""

    PUCK = "Puck"
    CHARON = "Charon"
    KORE = "Kore"
    FENRIR = "Fenrir"
    AOEDE = "Aoede"


class GeminiLiveResponseModality(StrEnum):
    """Response modalities accepted by Gemini Live."""

    AUDIO = "AUDIO"
    TEXT = "TEXT"


class GeminiLiveApiVersion(StrEnum):
    """Gemini API versions."""

    V1ALPHA = "v1alpha"
    V1BETA = "v1beta"
    V1 = "v1"


class GeminiLiveSampleRate(IntEnum):
    """Sample rates Gemini Live accepts on the input/output streams."""

    HZ_16000 = 16000
    HZ_24000 = 24000


class GeminiLiveEventType(StrEnum):
    """Adapter-level event-type strings yielded by :meth:`receive_events`."""

    AUDIO = "audio"
    TRANSCRIPT_INPUT = "transcript_input"
    TRANSCRIPT_OUTPUT = "transcript_output"
    FUNCTION_CALL = "function_call"
    RESPONSE_DONE = "response_done"
    SPEECH_STARTED = "speech_started"


# Default PCM audio format used on the wire for Gemini Live.
# Gemini Live requires PCM16 mono; sample-rate negotiation happens via
# ``speech_config``. Patter callers should resample to 16 kHz before calling
# :meth:`GeminiLiveAdapter.send_audio`.
DEFAULT_INPUT_SAMPLE_RATE_HZ = GeminiLiveSampleRate.HZ_16000.value
DEFAULT_OUTPUT_SAMPLE_RATE_HZ = GeminiLiveSampleRate.HZ_24000.value


class GeminiLiveAdapter:
    """Bridges a bidirectional audio stream to Google Gemini Live.

    The adapter presents the same surface as :class:`OpenAIRealtimeAdapter`:
    ``connect() -> send_audio(bytes) -> async for event in receive_events() -> close()``.

    Tool calling is supported via the ``tools`` constructor argument (same
    shape as the OpenAI adapter); function-call events are emitted as
    ``("function_call", {"call_id", "name", "arguments"})``.

    Requires ``google-genai>=1.55`` installed (``pip install getpatter[gemini-live]``).
    The SDK is imported lazily so callers that do not use Gemini Live do not
    pay the import cost.
    """

    def __init__(
        self,
        api_key: str,
        # gemini-2.0-flash-exp was experimental preview retired Dec 2024.
        # gemini-live-2.5-flash-preview was shut down Dec 9, 2025.
        # Current native-audio live model (v1alpha only) is the dated preview.
        # Override via GeminiLive(model=...) if needed.
        # TODO verify against Google docs: https://ai.google.dev/gemini-api/docs/live
        model: Union[
            GeminiLiveModel, str
        ] = GeminiLiveModel.NATIVE_AUDIO_PREVIEW_09_2025,
        voice: Union[GeminiLiveVoice, str] = GeminiLiveVoice.PUCK,
        instructions: str = "",
        language: str = "en-US",
        tools: list[dict] | None = None,
        input_sample_rate: Union[
            GeminiLiveSampleRate, int
        ] = GeminiLiveSampleRate.HZ_16000,
        output_sample_rate: Union[
            GeminiLiveSampleRate, int
        ] = GeminiLiveSampleRate.HZ_24000,
        temperature: float = 0.8,
    ) -> None:
        self.api_key = api_key
        self.model = model
        self.voice = voice
        self.instructions = instructions
        self.language = language
        self.tools = tools
        self.input_sample_rate = input_sample_rate
        self.output_sample_rate = output_sample_rate
        self.temperature = temperature
        self._client: Any = None
        self._session: Any = None
        self._session_cm: Any = None
        self._running = False
        # Tracks call_id -> function name so tool responses can be sent back
        # with the correct ``name`` field (Gemini expects the original function
        # name, not the call_id).
        self._pending_tool_calls: dict[str, str] = {}

    def __repr__(self) -> str:
        return (
            f"GeminiLiveAdapter(model={self.model!r}, voice={self.voice!r}, "
            f"language={self.language!r})"
        )

    async def connect(self) -> None:
        """Open a Live session with Gemini.

        Lazily imports ``google.genai`` — raises ``RuntimeError`` with a
        helpful install hint if the extra is missing.
        """
        try:
            from google import genai  # type: ignore[import-not-found]
            from google.genai import types as genai_types  # type: ignore[import-not-found]
        except ImportError as exc:  # pragma: no cover - exercised in production
            raise RuntimeError(
                "Gemini Live requires the 'google-genai' package. "
                "Install with: pip install getpatter[gemini-live]"
            ) from exc

        self._client = genai.Client(
            api_key=self.api_key,
            http_options={"api_version": GeminiLiveApiVersion.V1ALPHA.value},
        )

        speech_config = genai_types.SpeechConfig(
            voice_config=genai_types.VoiceConfig(
                prebuilt_voice_config=genai_types.PrebuiltVoiceConfig(
                    voice_name=self.voice,
                ),
            ),
            language_code=self.language,
        )

        config: dict[str, Any] = {
            "response_modalities": [GeminiLiveResponseModality.AUDIO.value],
            "speech_config": speech_config,
            "temperature": self.temperature,
            # Without these, native-audio sessions produced NO user
            # transcript ever and no assistant transcript in AUDIO modality —
            # logs/history/metrics got nothing for Gemini Live calls.
            "input_audio_transcription": {},
            "output_audio_transcription": {},
        }
        if self.instructions:
            config["system_instruction"] = genai_types.Content(
                parts=[genai_types.Part(text=self.instructions)],
            )
        if self.tools:
            from getpatter.providers.google_llm import _sanitize_gemini_schema

            config["tools"] = [
                {
                    "function_declarations": [
                        {
                            "name": t["name"],
                            "description": t.get("description", ""),
                            # Strip JSON-Schema keys the Live API's proto
                            # Schema rejects ($schema, additionalProperties —
                            # emitted by strict-mode and zod-derived MCP
                            # tools): one such tool 400'd the whole session.
                            "parameters": _sanitize_gemini_schema(
                                t.get("parameters", {})
                            ),
                        }
                        for t in self.tools
                    ],
                }
            ]

        # ``aio.live.connect`` returns an async context manager — we enter it
        # manually so that ``receive_events`` can stream indefinitely.
        self._session_cm = self._client.aio.live.connect(
            model=self.model,
            config=config,
        )
        self._session = await self._session_cm.__aenter__()
        self._running = True

    async def send_audio(self, audio: bytes) -> None:
        """Send a PCM16 mono chunk at ``input_sample_rate`` Hz."""
        if self._session is None:
            return
        mime_type = f"audio/pcm;rate={self.input_sample_rate}"
        await self._session.send_realtime_input(
            media={"data": audio, "mime_type": mime_type},
        )

    async def send_text(self, text: str) -> None:
        """Send a user text turn that triggers a spoken response."""
        if self._session is None:
            return
        await self._session.send_client_content(
            turns={"role": "user", "parts": [{"text": text}]},
            turn_complete=True,
        )

    async def send_function_result(self, call_id: str, result: str) -> None:
        """Return a tool call result to Gemini and continue the turn."""
        if self._session is None:
            return
        # Gemini requires the original function name in the response, not the
        # call_id. Look it up from the map populated when the tool call was
        # emitted; fall back to ``call_id`` if we never saw this id (defensive).
        name = self._pending_tool_calls.pop(call_id, call_id)
        await self._session.send_tool_response(
            function_responses=[
                {
                    "id": call_id,
                    "name": name,
                    "response": {"result": result},
                }
            ],
        )

    async def cancel_response(self) -> None:
        """Interrupt the current model turn (barge-in)."""
        # Gemini Live's ``send_realtime_input`` implicitly barges on VAD;
        # explicit cancel is not part of the v1alpha wire protocol.
        logger.debug("Gemini Live: cancel_response is implicit via VAD")

    async def receive_events(self) -> AsyncIterator[tuple[str, Any]]:
        """Yield ``(event_type, payload)`` tuples.

        Event types:
            ``audio`` — ``bytes`` of PCM16 mono audio at ``output_sample_rate``
            ``transcript_output`` — partial transcript text (if enabled)
            ``function_call`` — ``{"call_id", "name", "arguments"}``
            ``response_done`` — server indicated turn completion
        """
        if self._session is None:
            return

        try:
            async for response in self._session.receive():
                # ``response`` is a genai ``LiveServerMessage``. We gracefully
                # introspect it so the adapter stays stable across minor
                # google-genai releases.
                server_content = getattr(response, "server_content", None)
                if server_content is not None:
                    model_turn = getattr(server_content, "model_turn", None)
                    if model_turn is not None:
                        for part in getattr(model_turn, "parts", []) or []:
                            inline = getattr(part, "inline_data", None)
                            if inline is not None and getattr(inline, "data", None):
                                yield (GeminiLiveEventType.AUDIO.value, inline.data)
                            text = getattr(part, "text", None)
                            if text:
                                yield (
                                    GeminiLiveEventType.TRANSCRIPT_OUTPUT.value,
                                    text,
                                )
                    input_tx = getattr(server_content, "input_transcription", None)
                    if input_tx is not None and getattr(input_tx, "text", None):
                        yield (
                            GeminiLiveEventType.TRANSCRIPT_INPUT.value,
                            input_tx.text,
                        )
                    output_tx = getattr(server_content, "output_transcription", None)
                    if output_tx is not None and getattr(output_tx, "text", None):
                        yield (
                            GeminiLiveEventType.TRANSCRIPT_OUTPUT.value,
                            output_tx.text,
                        )
                    if getattr(server_content, "turn_complete", False):
                        yield (GeminiLiveEventType.RESPONSE_DONE.value, None)
                    if getattr(server_content, "interrupted", False):
                        yield (GeminiLiveEventType.SPEECH_STARTED.value, None)

                go_away = getattr(response, "go_away", None)
                if go_away is not None:
                    # Gemini Live hard-caps session length (~10-15 min without
                    # resumption); goAway is the only warning before the
                    # server drops the connection. Surface it loudly.
                    logger.warning(
                        "Gemini Live goAway received — session ends in %s",
                        getattr(go_away, "time_left", "unknown"),
                    )

                tool_call = getattr(response, "tool_call", None)
                if tool_call is not None:
                    for fn in getattr(tool_call, "function_calls", []) or []:
                        args = getattr(fn, "args", {}) or {}
                        call_id = getattr(fn, "id", "") or ""
                        fn_name = getattr(fn, "name", "") or ""
                        if call_id and fn_name:
                            self._pending_tool_calls[call_id] = fn_name
                        yield (
                            GeminiLiveEventType.FUNCTION_CALL.value,
                            {
                                "call_id": call_id,
                                "name": fn_name,
                                "arguments": json.dumps(args)
                                if not isinstance(args, str)
                                else args,
                            },
                        )
        except asyncio.CancelledError:
            raise
        except Exception as exc:  # pragma: no cover - defensive
            logger.error("Gemini Live receive error: %s", exc)
        finally:
            self._running = False

    async def close(self) -> None:
        """Tear down the Live session."""
        self._running = False
        if self._session_cm is not None:
            try:
                await self._session_cm.__aexit__(None, None, None)
            except Exception as exc:  # pragma: no cover
                logger.debug("Gemini Live close error: %s", exc)
            finally:
                self._session_cm = None
                self._session = None
        self._client = None
        self._pending_tool_calls.clear()
